### DISCUSSION TITLE
ローカルでコミット履歴を複数作り、pushする際にコミット履歴を1つに統合？する方法

### DISCUSSION THREAD

**Discussion by pss-aileen** [first-contributions-ja/first-contributions-ja.github.io#35] [10: 技術についての質問] [answered]
理解が足りないと感じたのでちょっと調査中。

自分の場合はこうしたら便利なのかな？というメモです。
- 変更を記録しておきたい場合、`git add .`、`git commit` までする（`push` しない）
  - ↑これは何度でもしても良い
- 最終的に `push`（みんなと共有する）前に、`git reset --mixed ハッシュ`でブランチ切る前直前のコミットに戻って、統合する
  - 挙動要確認→できた
- pushして、プルリクエストを出す→できた

### プルリクをDraftで出したり修正したりする場合 
- `git commit --amend --no-edit`、pushで修正できるか...？ここ調査必要
  - →push前まではできた
  - コミットまではできたが、git push origin HEAD だとできないので `git push -f origin HEAD`で、--amendでまとめたものがプッシュできた（必然的に上書きに..！？）

---

個人的にそこそこ心配性というか自分が書いたコードのログを何度でもとっておきたい（色々なパターンを見返すため）ので、どうするのが最適か探る

**Comment by pss-aileen** [comment, marked as answer]
@kagomen さん！質問です！

#22 こちらで一回、pushしたものを変更して再度pushされたかと思うのですが、それってどのようにされましたか？
やり方がわからなくて自分なりに以下のように試したのですが、同じような流れなのかが知りたいです！

- ブランチ作る、ブランチに移動
- ファイルの編集
- `git add .`
- `git commit -m "メッセージ"`
- `git push origin HEAD`
- ↑ここで `push` されるのでプルリクを出す
- 修正が入る
- ローカルで編集する
- `git commit --amend --no-edit` で直前のコミットにまとめる
- `git push -f origin HEAD` で強制的にpushして、プルリクエストにも反映される...

こんな感じでしょうか？

@kazzyfrog さん的にも、このような流れで作業されたりするものですか？

**Reply by kagomen** [reply]
同じ流れです！
自分の場合はコミットメッセージの変更があったので、`git commit --amend`でメッセージを編集しました！

**Comment by kazzyfrog** [comment]
これは、本当に色々な流派があると思います！
なので、そこら辺のルールを統一してもいいかもですね！

楽なのは、気にせずたくさんコミットして、Squash Mergeで、マージに１つのコミットにまとめる形ですかね。
そうすれば、mainブランチでは、１イシューに対して、対応する１コミットになります！

他には、mainではない、個人の作業用のブランチであれば、draftの状態（レビューを受ける前）なら、`push -f`でコミット整理するのもOKだと思います！
> @kazzyfrog さん的にも、このような流れで作業されたりするものですか？

なので、これで全然問題ないです！

**Reply by pss-aileen** [reply]
なるほど、流派があるのですね...！

Squashについてもちょっと勉強しなおします！

個人的にみんなどうやっているのかが気になったので質問してみました！ありがとうございます！

とりあえず自分は最初に書いた流れで今後は試してみようと思います！

**Comment by kazzyfrog** [comment]
@pss-aileen 
せっかくなので、よかったら #37 のPRでSquash Mergeしてみて下さい！
複数のコミットを1つにまとめられますよ🙌

**Reply by pss-aileen** [reply]
ありがとうございます！！やってみました！！
勉強になりました🙇

自分用メモ
https://qiita.com/ys-office-llc/items/201672d8e79c15aaad28
